server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,753 @@
+			buf += `<small>On an alternate account</small><br />`;
+		}
+		const lastLogin = this.getLastLogin(userid);
+		if (lastLogin && !connected) {
+			// THIS IS A TERRIBLE HACK BUT IT WORKS OKAY
+			const time = Chat.toTimestamp(new Date(Number(lastLogin)), {human: true});
+			buf += `Last login: ${time.split(' ').reverse().join(', on ')}`;
+			buf += ` (${Chat.toDurationString(Date.now() - lastLogin, {precision: 1})} ago)`;
+		}
+		buf = `<div class="infobox">${buf}</div>`;
+		return this.toLink(buf);
+	}
+	toLink(buf: string) {
+		return buf.replace(/<a roomid="/g, `<a target="replace" href="/`);
+	}
+	addFriend(senderID: User | string, receiverID: string) {
+		receiverID = toID(receiverID);
+		senderID = toID(senderID);
+		if (!receiverID) throw new Chat.ErrorMessage(`Invalid user.`);
+
+		if (this.hasFriendship(senderID, receiverID)) {
+			throw new Chat.ErrorMessage(`You and ${receiverID} are already friends.`);
+		}
+		const sender = Users.getExact(senderID);
+		const receiver = Users.getExact(receiverID);
+		if (sender) {
+			if (!sender.friends) sender.friends = new Set();
+			sender.friends.add(receiverID);
+		}
+		if (receiver?.friendRequests) {
+			if (!receiver.friends) receiver.friends = new Set();
+			receiver.friends.add(senderID);
+		}
+
+		this.statements.add.run({userid: senderID, friend: receiverID, login: Date.now()});
+		this.statements.add.run({userid: receiverID, friend: senderID, login: Date.now()});
+	}
+	removeFriend(userid: string, friendID: string) {
+		userid = toID(userid);
+		friendID = toID(friendID);
+		if (!friendID || !userid) throw new Chat.ErrorMessage(`Invalid usernames supplied.`);
+
+		const user = Users.getExact(userid);
+		const friend = Users.getExact(friendID);
+		if (user) {
+			if (!user.friends) user.friends = new Set();
+			if (!user.friends.has(friendID)) throw new Chat.ErrorMessage(`You do not have '${friendID}' friended.`);
+			user.friends.delete(friendID);
+		}
+		if (friend) {
+			if (!friend.friends) friend.friends = new Set();
+			friend.friends.delete(userid);
+		}
+
+		this.statements.delete.run({userid});
+	}
+	notifyPending(user: User) {
+		if (user.settings.blockFriendRequests) return;
+		const pendingCount = user.friendRequests?.received.size;
+		if (!pendingCount || pendingCount < 1) return;
+		const prefix = `|pm|&|${user.getIdentity()}|`;
+		user.send(`${prefix}You have ${pendingCount} friend requests pending!`);
+		user.send(`${prefix}/raw <button class="button" name="send" value="/j view-friends-received">View</button></div>`);
+	}
+	notifyConnection(user: User) {
+		const message = `/nonotify Your friend ${Utils.escapeHTML(user.name)} has just connected!`;
+		for (const userid of user.friends!) {
+			const curUser = Users.get(userid);
+			if (curUser?.settings.allowFriendNotifications) {
+				curUser.send(`|pm| &|${curUser.getIdentity()}|${message}`);
+			}
+		}
+	}
+	hasFriendship(userid1: string, userid2: string) {
+		userid1 = toID(userid1);
+		userid2 = toID(userid2);
```

Comment:
Why not just take IDs as parameters, instead of coercing them? Again, another situation where parameters are either misnamed or redundant